lib/instrumentation/grpc-js/grpc.js

```diff
@@ -6,80 +6,215 @@
-            nrMetadata.add(key, outboundAgentHeaders[key])
-          })
-        } else {
-          shim.logger.debug('Distributed tracing disabled by instrumentation.')
+        // Java captures client errors based on status code
+        // but has specific gRPC configuration to turn off
+        if (code !== 0) {
+          // this is currently just creating an error from the details string
+          shim.agent.errors.add(segment.transaction, details)
         }
 
-        args[0] = nrMetadata
+        segment.addAttribute('component', 'gRPC')
 
-        const originalListener = args[1]
-        const nrListener = Object.assign({}, originalListener)
-        nrListener.onReceiveStatus = (status) => {
-          const { code, details } = status
+        const protocol = 'grpc'
 
-          segment.addAttribute('grpc.statusCode', code)
-          segment.addAttribute('grpc.statusText', details)
+        const url = `${protocol}://${authorityName}${this.methodName}`
 
-          // Java captures client errors based on status code
-          // but has specific gRPC configuration to turn off
-          if (code !== 0) {
-            // this is currently just creating an error from the details string
-            shim.agent.errors.add(segment.transaction, details)
-          }
+        segment.addAttribute('http.url', url)
+        segment.addAttribute('http.method', this.methodName)
 
-          segment.addAttribute('component', 'gRPC')
+        segment.end()
 
-          const protocol = 'grpc'
+        originalListener && originalListener.onReceiveStatus(status)
+      }
 
-          const url = `${protocol}://${authorityName}${this.methodName}`
+      args[1] = nrListener
 
-          segment.addAttribute('http.url', url)
-          segment.addAttribute('http.method', this.methodName)
+      return original.apply(this, args)
+    }
+  }
+}
 
-          segment.end()
+/**
+ * Instruments the grpc-js server by intercepting the moment when
+ * server methods are registered from the method implementations
+ * provided to grpc-js. This handles all four types of server
+ * invocations: unary, client-streaming, server-streaming, and
+ * bidirectional streaming.
+ *
+ * @param {object} shim the web shim to instrument with
+ * @param {Function} original the original function
+ * @returns {Function} the instrumented function
+ */
+function wrapRegister(shim, original) {
+  const constants = shim.require('./build/src/constants')
 
-          originalListener && originalListener.onReceiveStatus(status)
-        }
+  return function wrappedRegister() {
+    const args = shim.argsToArray.apply(shim, arguments)
 
-        args[1] = nrListener
+    const name = args[0]
+    const handler = args[1]
+    const type = args[args.length - 1]
 
-        return original.apply(this, args)
-      }
+    if (this.handlers.has(name)) {
+      shim.logger.debug(
+        `not instrumenting again gRPC method handler for ${name}: it is already registered in the server`
```

Comment:
```suggestion
        `Not instrumenting gRPC method handler for ${name}: it is already registered in the server.
```

Reply:
Hm? Why remove the `again`? It should have been instrumented the first time; no need to do it again.

Reply:
maybe i misunderstood the intent here.  but there's some redundancy going on.  `instrumenting again gRPC method handler` and `it is already registered in the server`.  If this.handlers has the name is means we aren't going to instrument the handler again correct?

Reply:
Maybe I should just omit saying anything. The call is `server.register('lol', lolHandler)` once and we instrument it and it's registered. You call `server.register('lol', lolHandler)` again and that means we aren't doing any instrumenting, and we pass control back to `server.register` that also does nothing because it sees it already has an (instrumented) implementation for `'lol'`.

Maybe none of this drama is worth logging about.